src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,170 @@
+#include <logic.hpp>
+#include <math.hpp>
+#include <reduce.hpp>
+#include <select.hpp>
+#include <svd.hpp>
+#include <transpose.hpp>
+
+using af::dim4;
+using af::dtype_traits;
+using std::vector;
+using std::swap;
+
+using namespace detail;
+
+const double dfltTol = 1e-6;
+
+template<typename T>
+Array<T> pinverseSvd(const Array<T> &in, const double tol)
+{
+    // Moore-Penrose Pseudoinverse
+
+    in.eval();
+    int M = in.dims()[0];
+    int N = in.dims()[1];
+
+    // Compute SVD
+    typedef typename dtype_traits<T>::base_type Tr;
+    Array<Tr> sVec = createEmptyArray<Tr>(dim4(min(M, N)));
+    Array<T> u = createEmptyArray<T>(dim4(M, M));
+    Array<T> vT = createEmptyArray<T>(dim4(N, N));
+    svd<T, Tr>(sVec, u, vT, in);
+
+    // Cast s back to original data type for matmul later
+    // (since svd() makes s' type the base type of T)
+    Array<T> sVecCast = cast<T, Tr>(sVec);
+
+    Array<T> v = transpose(vT, true);
+
+    // Get reciprocal of sVec's non-zero values for s pinverse, except for
+    // very small non-zero values though (< relTol), in order to avoid very
+    // large reciprocals
+    double relTol = tol * static_cast<double>(max(M, N))
+                        * reduce_all<af_max_t, Tr, Tr>(sVec);
+    Array<T> relTolArr = createValueArray<T>(sVecCast.dims(), scalar<T>(relTol));
+    Array<T> ones = createValueArray<T>(sVecCast.dims(), scalar<T>(1.));
+    Array<T> sVecRecip = arithOp<T, af_div_t>(ones, sVecCast, sVecCast.dims());
+    Array<char> cond = logicOp<T, af_ge_t>(sVecCast, relTolArr,
+                                           sVecCast.dims());
+    Array<T> zeros = createValueArray<T>(sVecCast.dims(), scalar<T>(0.));
+    sVecRecip = createSelectNode<T>(cond, sVecRecip, zeros, sVecRecip.dims());
+    sVecRecip.eval();
+
+    // Make s vector into s pinverse array
+    Array<T> sPinv = diagCreate<T>(sVecRecip, 0);
+
+    Array<T> uT = transpose(u, true);
+
+    // Crop v and u* for final matmul later based on s+'s size, because
+    // sVec produced by svd() has minimal dim length (no extra zeroes).
+    // Thus s+ produced by diagCreate() will have minimal dims as well,
+    // and v could have an extra dim0 or u* could have an extra dim1
+    if (v.dims()[1] > sPinv.dims()[0]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(v.dims()[0] - 1), 1.},
+            {0., static_cast<double>(sPinv.dims()[0] - 1), 1.}
+        };
+        v = createSubArray<T>(v, seqs);
+    }
+    if (uT.dims()[0] > sPinv.dims()[1]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(sPinv.dims()[1] - 1), 1.},
+            {0., static_cast<double>(uT.dims()[1] - 1), 1.}
+        };
+        uT = createSubArray<T>(uT, seqs);
+    }
+
+    Array<T> out = matmul<T>(matmul<T>(v, sPinv, AF_MAT_NONE, AF_MAT_NONE),
+                             uT, AF_MAT_NONE, AF_MAT_NONE);
+
+    return out;
+}
+
+template<typename T>
+static inline af_array pinverse(const af_array in, const double tol)
```

Comment:
I think you should think about moving the batching logic into the pinverseSvd function. You should use this intermediate function to convert from af_array to Array<T> objects. This makes the internal API more usable in other functions because it has the same behavior as the public API. 

So in this case this pinverse function should look like this:

```
template<typename T>
af_array pinverse(const af_array in, const double tol) {
  return getHandle(pinverse(getHandle<T>(in), tol)));
}
```

Reply:
Ah I understand, yep that makes sense let me move the batching logic indeed.